TLS 1.3 server-side mutual authentication implementation

ryru opened the pull request:
This PR covers:

- #3

Commit:
tls-crypto: Move AEAD ownership to the protection layer

This separates key derivation from key switching.

Commit:
tls-server: Mutual authentication support for TLS 1.3

This commit also addresses the side effect that additional messages has
an influence on the derivation of the application key secrets. Therefore
app key derivation is relocated after the server finished message to
derive the keys by using the correct set of input messages.

Commit:
tls-test: Add server-side mutual authentication support.

Commit:
tls-server: Generalise trusted public key search

Client and server share the same code which is relocated in the server
now and forward declared in the client.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1589,6 +1587,12 @@ METHOD(tls_handshake_t, build, status_t,
+					this->alert->add(this->alert, TLS_FATAL, TLS_INTERNAL_ERROR);
+					return NEED_MORE;
+				}
+				this->crypto->change_cipher(this->crypto, FALSE);
```
Wouldn't it be enough to just move this call here and do the key derivation as before with a single call?

ryru replied:
In my understanding no. `derive_labeled_keys()` previously derived inbound and outbound keys and also set it to in both directions `this->aead_in` and `this->aead_out`.
It may be possible to combine `this->crypto->derive_app_keys()` and `this->crypto->change_cipher()` to a single call though.

tobiasbrunner replied:
You're right. The keys in the existing `tls_aead_t` objects were set, which immediately changed the keys that are in use (i.e. calling `change_cipher()` is redundant after the first call). I guess this separation was necessary for the change_cipher_spec use case and the keys were also never changed in older versions. So for TLS 1.3, we could theoretically just call `change_cipher()` in `derive_handshake_keys()` and be done with it.

However, it might be cleaner to actually separate the key derivation from enabling a certain set of keys on the TLS protection layer. I guess the ownership of the `tls_aead_t` object would then be transferred to the protection layer whenever the method is called. New objects would later be created when a new set of keys is derived. That way the key derivation and the switch to new keys could both be done when appropriate.

ryru replied:
Let's discuss in today's meeting.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2132,16 +2159,36 @@ static bool derive_labeled_key(private_tls_crypto_t *this, bool server,
 	{
-		DBG1(DBG_TLS, "deriving key material failed");
-		goto out;
+		get_handshake_until(TLS_FINISHED, this->handshake, &hs);
```
Shouldn't it be possible to avoid this by doing the key derivation at the right time?

ryru replied:
No, I think in the current code not. Key derivation also sets keys on `this->aead_in` respectively `this->aead_out` in `derive_labled_key()`:

```
	if (!aead->set_keys(aead, chunk_empty, key, iv))
	{
		DBG1(DBG_TLS, "setting AEAD key material failed");
		goto out;
	}
	success = TRUE;
```

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -603,30 +603,7 @@ static status_t process_certificate(private_tls_peer_t *this,
-		enumerator->destroy(enumerator);
-	}
-	return public;
+	return tls_find_public_key(this->server_auth);
```
Why not replace `find_public_key()` completely?

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -744,11 +738,6 @@ Suite *socket_suite_create()
 	add_tls_test(test_tls13, TLS_1_3);
 	suite_add_tcase(s, tc);
 
-	tc = tcase_create("TLS 1.3/mutl");
-	tcase_add_checked_fixture(tc, setup_creds, teardown_creds);
-	add_tls_test(test_tls13_mutual, TLS_1_3);
-	suite_add_tcase(s, tc);
-
```
Why did you remove this? (Interesting that it worked before, I guess the test does not actually enforce any mutual authentication.)

ryru replied:
Until now the server never sent a _CertificateRequest_ message in TLS 1.3 test cases. With the server implementation it does now but the client is not ready now. I add the test again once the client is implemented.

tobiasbrunner replied:
I see.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -859,6 +879,14 @@ static status_t process_key_exchange(private_tls_server_t *this,
+{
+	return tls_find_public_key(this->peer_auth);
+}
+
```
Same as on the client, why not call `tls_find_public_key()` directly?

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -870,43 +898,70 @@ static status_t process_cert_verify(private_tls_server_t *this,
-		sig->destroy(sig);
-		if (verified)
+		enumerator = lib->credmgr->create_public_enumerator(lib->credmgr,
+								KEY_ANY, this->peer, this->peer_auth, TRUE);
```
I think you could just use the same function that does the lookup via certificate for all TLS versions instead of looping through all keys here (not sure if that was ever necessary).

ryru replied:
I add an implementation suggestion for this.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1359,6 +1413,55 @@ static status_t send_certificate_request(private_tls_server_t *this,
+		writer->write_uint8(writer, 0);
+
+		extensions = bio_writer_create(32);
+		DBG2(DBG_TLS, "sending extension within certificate_reqst message: %N",
```
`certificate_reqst` looks a bit weird. Not sure if a message is even necessary.

ryru replied:
I shortened the message to `sending CertificateRequest extension: %N` this is also more similar to the extension DBG2 messages sending in the ServerHello message. Do you agree?

tobiasbrunner replied:
I'd make it the same as the others, i.e. remove `CertificateRequest`, because we see `sending TLS CertificateRequest handshake` in the log right afterwards (well, after the DNs for the CAs - adding them before the other extensions might clarify this even further).

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1581,6 +1684,11 @@ METHOD(tls_handshake_t, build, status_t,
 			case STATE_CIPHERSPEC_CHANGED_OUT:
 				return send_encrypted_extensions(this, type, writer);
 			case STATE_ENCRYPTED_EXTENSIONS_SENT:
+				if (this->peer)
```
This doesn't seem to be the case for older TLS versions.

ryru replied:
Right, this only triggers a CertificateRequest message in TLS 1.3 but not in older TLS versions. Do I understand you correctly that such a functionality is desirable also for the TLS legacy branch?

tobiasbrunner replied:
That's what I'm wondering. It seems a certificate request was always sent with older versions, no matter if a peer identity was configured (which, according to the comment, was intended to enable this - if none is set, the client's authentication and certificate is verified but is allowed to fail, not sure what the purpose of that would be, though). So we perhaps should not make this conditional here either, or also add such a check for older versions. It really depends on whether it still works for our use cases if it's made conditional.

ryru replied:
Yes, this is the case. In older TLS versions the server always sent a _CertificateRequest_ message. I see three options:

a) I make it conditional in legacy TLS: This results in the situation that we currently not know, if it has an impact on how libtls is used in strongSwan.
b) I remove the conditional in TLS 1.3: This results in changes in the `tls_peer.c` because it receives a _CertificateRequest_ message but has not yet an implementation for mutual authentication.
c) leave as is: This does not break the old implementation an does not touch `tls_peer.c` in this commit.

I would suggest to leave it as is for now. Once the TLS 1.3 mutual authentication is done, I remove the conditional in a separate commit to get consistency. Do you agree?

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1581,6 +1684,11 @@ METHOD(tls_handshake_t, build, status_t,
+				if (this->peer)
+				{
+					return send_certificate_request(this, type, writer);
+				}
+			case STATE_CERTREQ_SENT:
```
Is this fall-through on purpose? If so, add a comment to clarify.

tobiasbrunner commented on scripts/tls_test.c:
```diff
@@ -402,7 +430,15 @@ int main(int argc, char *argv[])
-		res = serve(host, server, times, cache, min_version, max_version);
+		if (cauth)
+		{
+			client = find_client_id();
```
Not sure if this makes sense as the client certificate won't be loaded on the server, i.e. that function will only find the server identity. Maybe just assign ID_ANY (`identification_create_from_encoding(ID_ANY, chunk_empty);`).  You also wouldn't need the `cauth` flag. Instead assign that identity to `client`, which you initialize to `NULL`, above in the arg parse loop.

ryru replied:
Thanks, this looks better.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -879,46 +879,27 @@ static status_t process_key_exchange(private_tls_server_t *this,
+	{
+		if (!this->crypto->verify_handshake(this->crypto, public, reader))
 		{
 			DBG1(DBG_TLS, "no trusted certificate found for '%Y' to verify TLS peer",
```
This error message is not correct anymore. Should probably be the same as below for TLS 1.3. The alert should also be `TLS_DECRYPT_ERROR` in both TLS versions (see section 4.4.3). And there is a `public->destroy(public);` call required there before returning here. But shouldn't it be possible to combine the two paths?

ryru replied:
I addressed your comments.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -879,46 +879,27 @@ static status_t process_key_exchange(private_tls_server_t *this,
-			DBG1(DBG_TLS, "signature verification failed, trying another key");
-		}
-		enumerator->destroy(enumerator);
+		DBG1(DBG_TLS, "no TLS public key found for peer '%Y'", this->peer);
+		this->alert->add(this->alert, TLS_FATAL, TLS_CERTIFICATE_UNKNOWN);
+		return NEED_MORE;
```
The `peer_auth_optional` use case must also be covered here.

ryru replied:
I think I addressed this. I do not see the purpose of `peer_auth_optional` though.

tobiasbrunner commented on scripts/tls_test.c:
```diff
@@ -305,6 +325,7 @@ int main(int argc, char *argv[])
 			{"port",		required_argument,		NULL,		'p' },
 			{"cert",		required_argument,		NULL,		'x' },
 			{"key",			required_argument,		NULL,		'k' },
+			{"cafile",		required_argument,		NULL,		'f' },
```
Better call this `cacert`.

Commit:
tls-server: Make CertificateRequest conditional in old TLS versions

The server implementation now only sends a CertificateRequest message if
it has identity information to verify client certificates.

Commit:
tls-server: Terminate connection if peer certificate is required but not sent

This change mainly affects legacy TLS versions because TLS 1.3
connections are terminated by the server once the peer does not send a
CertificateVerify message next to its empty Certificate message.

Commit:
tls-hkdf: Use base secret instead of traffic secret to derive finish

to verify handshake authentication the secret has to be derived from the
base secret instead of the latest handshake state.

tobiasbrunner commented on src/libtls/tls_hkdf.c:
```diff
@@ -67,11 +67,21 @@ struct private_tls_hkdf_t {
+	 * proper finished key.
+	 */
+	chunk_t client_base_traffic_secret;
+	chunk_t server_base_traffic_secret;
```
The names seem not ideal as there are different base keys throughout. Maybe call them what they actually are, `*_handshake_traffic_secret`.

Also, it might be preferable to store these a bit differently, so they can be passed around (also see below). For instance, define something like `struct cached_secrets_t { chunk_t client; chunk_t server; }` so you can then define these as `cached_secrets_t handshake_traffic_secrets;` and `cached_secrets_t traffic_secrets;`.

tobiasbrunner commented on src/libtls/tls_hkdf.c:
```diff
@@ -482,6 +496,11 @@ static bool get_shared_label_keys(private_tls_hkdf_t *this, chunk_t label,
+	{
+		secret = server ? this->server_base_traffic_secret
+						: this->client_base_traffic_secret;
+	}
```
Instead of this, pass a `cached_secrets_t*` to this helper (i.e. `&this->handshake_traffic_secrets` from `derive_finished`, `&this->traffic_secrets` from the others).

tobiasbrunner commented on src/libtls/tls_hkdf.c:
```diff
@@ -446,12 +456,16 @@ METHOD(tls_hkdf_t, generate_secret, bool,
 	{
 		case TLS_HKDF_C_HS_TRAFFIC:
+			chunk_clear(&this->client_base_traffic_secret);
+			this->client_base_traffic_secret = chunk_clone(okm);
```
Add a comment after this to clarify the fall-through (same below).

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2164,8 +2164,17 @@ static bool derive_labeled_keys(private_tls_crypto_t *this,
 								tls_hkdf_label_t server_label)
 {
-	tls_aead_t *aead_c = this->aead_out, *aead_s = this->aead_in;
+	tls_aead_t *aead_c, *aead_s;
```
Attaching this here, but it's really about the commit message. It describes that keys can be derived separately, but that's not the case. It's more that it allows separating key derivation from key switching if the derivation does not modify the  `tls_aead_t` objects currently in use by the protection layer.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1561,7 +1550,10 @@ METHOD(tls_handshake_t, build, status_t,
+				if (this->peer)
+				{
+					return send_certificate_request(this, type, writer);
+				}
```
Add a comment regarding the fall-through.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -86,11 +86,6 @@ struct private_tls_server_t {
-	/**
-	 * Is it acceptable if we couldn't verify the peer certificate?
-	 */
-	bool peer_auth_optional;
```
In this patch, there should be additional removed code that used this flag (in `process_cert_verify()`). Maybe try editing this commit via `rebase -i`. You'll get some compile errors if you try to build this commit, so fix the code and amend the commit. There will then probably be some conflicts with the mutual auth commit when continuing the rebase, but you can  just apply all changes there as it does not use the flag anymore.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -671,6 +671,12 @@ static status_t process_certificate(private_tls_server_t *this,
 	certs = bio_reader_create(data);
+	if (!certs->remaining(certs))
+	{
+		DBG1(DBG_TLS, "no cert sent by peer");
```
Better write `certificate` instead of `cert`. But I wonder whether this is actually necessary. If the client did not send a certificate, shouldn't `tls_find_public_key()` fail? (It requires a certificate in peer_auth.)

ryru replied:
If the client does not have any certificate to send it sends a _Certificate_ message without any certificate in it (certificate length == 0). In TLS 1.3 the connection is terminated by the server when no _CertificateVerify_ message is present.

In legacy TLS `process_cert_verify()`, which calls `tls_find_public_key()`, is never called because a client sends a _ChangeCipherSpec_ message between its _KeyExchange_ and _Finished_ messages. tls_fragmentation.c sets status to `FAILED` in the `process()` method since the `cipherspec_changed()` evaluates FALSE because `this->peer` is not null and the current state is `STATE_KEY_EXCHANGE_RECEIVED` and not `STATE_CERT_VERIFY_RECEIVED`. This means if a TLS 1.2 client tries to open a connection without a client certificate the handshake hangs after server output `processing TLS ChangeCipherSpec record (1 bytes)` which does not look right to me. The above change terminates the connection to the client.

ryru replied:
As discussed in today's meeting, we leave the current code since it is functional but open a new issue to investigate the case that the connection hangs after a _ChangeCipherSpec_ message.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -940,14 +940,7 @@ static status_t process_finished(private_tls_server_t *this,
 			return NEED_MORE;
 		}
-
-		if (!this->crypto->derive_app_keys(this->crypto))
```
I'd add some notes on the reasoning behind moving this code to the commit message. Because the side effect the additional messages sent by the client have on it might not be immediately obvious.

tobiasbrunner replied:
The commit message looks weird, like all text is on a single line. I'd also add an "also" (i.e. "This commit **also** addresses") because it does this in addition to adding mutual auth support for TLS 1.3 (which might actually be something you could add to the first line of the message, because mutual auth is already supported for older versions).

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1580,12 +1635,25 @@ METHOD(tls_handshake_t, build, status_t,
+					this->alert->add(this->alert, TLS_FATAL, TLS_INTERNAL_ERROR);
+					return NEED_MORE;
+				}
+				this->crypto->change_cipher(this->crypto, FALSE);
```
Maybe add a comment that explains that the inbound key is switched after processing the client's Finished message.

tobiasbrunner commented on src/libtls/tls_hkdf.c:
```diff
@@ -67,11 +72,14 @@ struct private_tls_hkdf_t {
+	cached_secrets_t handshake_traffic_secrets;
+
+	/**
+	 * Application traffic secrets.
```
Not exclusively used for the application traffic secrets, initially also stores a copy of the handshake traffic secrets. So maybe use "Current" instead of "Application".

tobiasbrunner commented on src/libtls/tls_hkdf.c:
```diff
@@ -446,16 +454,22 @@ METHOD(tls_hkdf_t, generate_secret, bool,
 		case TLS_HKDF_C_HS_TRAFFIC:
+			chunk_clear(&this->handshake_traffic_secrets.client);
+			this->handshake_traffic_secrets.client = chunk_clone(okm);
+			/* otherwise fall through to next state */
```
```suggestion
			/* fall-through */
```
That's usually enough if no explicit explanation for the fall-through is needed.

tobiasbrunner commented on src/libtls/tls_hkdf.c:
```diff
@@ -501,25 +515,34 @@ static bool get_shared_label_keys(private_tls_hkdf_t *this, chunk_t label,
+	chunk_clear(&secrets->client);
+	chunk_clear(&secrets->server);
+}
+
```
I'd move this either near the struct declaration, or closer to where it's actually used (e.g. right above the `destroy` method).

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2555,3 +2555,34 @@ enumerator_t *tls_get_supported_key_types(tls_version_t min_version,
 	return enumerator_create_filter(ht->create_enumerator(ht),
 									filter_key_types, ht, destroy_key_types);
 }
+
```
As discussed, since the function does not depend on anything in `tls_crypto.c`, it might be preferable to simply rename the function and make it non-static so it can be called from the other file via forward declaration (similar to `tls_write_key_share()`). But maybe move it to `tls_server.c` so all forward declarations between these two files are in `tls_peer.c`.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -860,43 +880,28 @@ static status_t process_key_exchange(private_tls_server_t *this,
-			break;
-		}
-		DBG1(DBG_TLS, "signature verification failed, trying another key");
+		DBG1(DBG_TLS, "no TLS public key found for peer '%Y'", this->peer);
```
I wonder if the log message in the previous code was clearer for users. For one, "TLS public key" is not really a thing, and "no trusted certificate" might describe the issue better. You could perhaps change this similarly for the client in the `tls_find_public_key()` commit.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1350,6 +1359,57 @@ static status_t send_certificate_request(private_tls_server_t *this,
+		/* certificate request context as described in RFC 8446, section 4.3.2 */
+		writer->write_uint8(writer, 0);
+
+		authorities = bio_writer_create(64);
+		write_certificate_authorities(authorities);
```
I'd move this down, similar to how you did that with the other extension and its `supported` writer.

ryru replied:
Do you mean moving these two lines:

```
	authorities = bio_writer_create(64);
	write_certificate_authorities(authorities);
```

between the two extensions?

I think I moved this on purpose because of your comment a week ago:

> I'd make it the same as the others, i.e. remove CertificateRequest, because we see sending TLS CertificateRequest handshake in the log right afterwards (well, after the DNs for the CAs - adding them before the other extensions might clarify this even further).

Did I misunderstand you?

tobiasbrunner replied:
I see, sorry about that. The code is just a bit strange this ways and the log messages are too (sending cert reqs, followed by the signature_algorithms extension and only then the certificate_authorities extension). I'd switch the order of the two extensions, so creation and use of the writer are closer together and the log would make a bit more sense (maybe if the cert_authorities extension is sent first, actually moving these lines after the "sending extension" log message would also be OK).

ryru replied:
I addressed this comment in the latest commits.

tobiasbrunner commented on scripts/tls_test.c:
```diff
@@ -38,7 +38,27 @@ static void usage(FILE *out, char *cmd)
+	fprintf(out, "  --cert <file>                certificate to use to establish and verify TLS connection\n");
+	fprintf(out, "  --key <file>                 private key to use to establish and verify TLS connection\n");
+	fprintf(out, "  --cacert <file>              certificate to use to authenticate clients.\n");
+	fprintf(out, "                               This configuration enables required client authentication\n");
```
As discussed, clarify the use of these arguments (`--cacert` to verify peer, the others to authenticate itself).

tobiasbrunner commented on scripts/tls_test.c:
```diff
@@ -38,7 +38,27 @@ static void usage(FILE *out, char *cmd)
-	fprintf(out, "  %s --listen <address> --port <port> --key <key> [--cert <file>]+ [--times <n>]\n", cmd);
+	fprintf(out, "  %s --listen <address> --port <port> --key <key> [--cert <file>]+ [--cacert <file>]+ [--times <n>]\n", cmd);
+
+	fprintf(out, "\n\nOptions:\n");
```
`pki` and `swanctl` both don't add a separator between usage lines and the option descriptions (which are indented more, though - but strangely not the same, `pki` indents with 8 spaces, `swanctl` with 11, not sure why, or why both indent the options that much at all, even without separator, less should be enough). I'm not opposed to adding a separator and, in particular, indenting the options less. But maybe make `options:` lower case (to match `usage:`, which is lower case in the other tools too) with only a single empty line before, or use a single empty line as separator and indent the options by 4 spaces (and reduce the space between options and their descriptions a bit).

tobiasbrunner commented on scripts/tls_test.c:
```diff
@@ -37,8 +37,27 @@
+	fprintf(out, "  %s --connect <address> --port <port> [--key <key] [--cert <file>] [--cacert <file>]+ [--times <n>]\n", cmd);
+	fprintf(out, "  %s --listen <address> --port <port> --key <key> --cert <file> [--cacert <file>]+ [--times <n>]\n", cmd);
+	fprintf(out, "\n");
+	fprintf(out, "options:\n");
+	fprintf(out, "    --help                   print help and exit\n");
```
Not sure, but you might have overlooked the "or" in my comment about this :) With the "options:" separator, using two spaces indentation should be perfectly fine (and even provides two characters more for the description - I might adapt the other tools to this too because using that much indentation there seems wasteful). The four space indentation was a suggestion in case the options were separated by an empty line only, so they can be better distinguished from the usage lines that are indented by two spaces.

tobiasbrunner reviewed (approved):
Just a small change request in `tls_test.c`, otherwise looks good.

ryru commented:
Thanks for the feedback.
